exec/java-exec/src/main/java/org/apache/drill/exec/store/pcapng/PcapngRecordReader.java

```diff
@@ -0,0 +1,218 @@
+
+  private final Path pathToFile;
+  private OutputMutator output;
+  private ImmutableList<ProjectedColumnInfo> projectedCols;
+  private FileSystem fs;
+  private FSDataInputStream in;
+  private List<SchemaPath> columns;
+
+  private Iterator<IPcapngType> it;
+
+  public PcapngRecordReader(final String pathToFile,
+                            final FileSystem fileSystem,
+                            final List<SchemaPath> columns) {
+    this.fs = fileSystem;
+    this.pathToFile = fs.makeQualified(new Path(pathToFile));
+    this.columns = columns;
+    setColumns(columns);
+  }
+
+  @Override
+  public void setup(final OperatorContext context, final OutputMutator output) throws ExecutionSetupException {
+    try {
+
+      this.output = output;
+      this.in = fs.open(pathToFile);
+      PcapDecoder decoder = new PcapDecoder(IOUtils.toByteArray(in));
+      decoder.decode();
+      this.it = decoder.getSectionList().iterator();
+      setupProjection();
+    } catch (IOException io) {
+      throw UserException.dataReadError(io)
+          .addContext("File name:", pathToFile.toUri().getPath())
+          .build(logger);
+    }
+  }
+
+  @Override
+  public int next() {
+    if (isSkipQuery()) {
+      return getBatchOfBlocks().size();
+    } else {
+      return putToTable(getBatchOfBlocks());
+    }
+  }
+
+  private int putToTable(final List<IEnhancedPacketBLock> batchOfBlocks) {
+    int counter = 0;
+    for (IEnhancedPacketBLock bLock : batchOfBlocks) {
+      for (ProjectedColumnInfo pci : projectedCols) {
+        pci.getColumn().process(bLock, pci.getVv(), counter);
+      }
+      counter++;
+    }
+    return counter;
+  }
+
+  @Override
+  public void close() throws Exception {
+    if (in != null) {
+      in.close();
+    }
+  }
+
+  private void setupProjection() {
+    if (isSkipQuery()) {
+      projectedCols = projectNone();
+    } else if (isStarQuery()) {
+      projectedCols = projectAllCols(Schema.getColumnsNames());
+    } else {
+      projectedCols = projectCols(columns);
+    }
+  }
+
+  private ImmutableList<ProjectedColumnInfo> projectNone() {
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+    pciBuilder.add(makeColumn("dummy", new DummyImpl()));
+    return pciBuilder.build();
+  }
+
+  private ImmutableList<ProjectedColumnInfo> projectAllCols(final Set<String> columns) {
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+    for (String colName : columns) {
+      pciBuilder.add(makeColumn(colName, Schema.getColumns().get(colName)));
+    }
+    return pciBuilder.build();
+  }
+
+  private ImmutableList<ProjectedColumnInfo> projectCols(final List<SchemaPath> columns) {
+    ImmutableList.Builder<ProjectedColumnInfo> pciBuilder = ImmutableList.builder();
+    for (SchemaPath schemaPath : columns) {
+      String projectedName = schemaPath.rootName();
+      if (Schema.getColumns().containsKey(projectedName.toLowerCase())) {
```

Comment:
This code assumes that all columns are matched. Actually, if the user requests the `foo` column which does not exist in the pcapng file, the proper behavior is, as described above, to create a dummy nullable int column, with the requested name.

Said another way, if the user asks for three columns, this reader must provide three columns, even if it has to "make up" nullable int columns to do so.

Looks like the code intended to handle this case: there is a `DummyImpl` that seems like what is needed here if there is no match for the column name.

Reply:
This works because Drill sets to all of the columns a MajorType as optional before initialization in a reader.